### PR TITLE
[@types/pdfmake] table layout & widths errors #43758

### PR DESCRIPTION
**Pull request by syffs**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition: changes related to #43758
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    - tableLayouts [support user-defined layouts](https://pdfmake.github.io/docs/document-definition-object/tables/)
    - table.widths [can be a string](https://github.com/bpampuch/pdfmake/blob/35f3ea19918f7a3b04aa54ad25e94db0c1bd5dac/src/DocMeasure.js#L634)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/pdfmake/interfaces.d.ts [hunk 1]
```diff
@@ -108,7 +108,7 @@ export type TableCell =
 
 export interface Table {
     body: TableCell[][];
-    widths?: Size[];
+    widths?: string | Size[];
```
Comment (jeralm) [inline types/pdfmake/interfaces.d.ts:111]: Can you provide a reference? I have not been able to find anything corroborating this.

File: types/pdfmake/interfaces.d.ts [hunk 2]
```diff
@@ -453,7 +453,7 @@ export interface ContextPageSize {
 export interface BufferOptions {
     fontLayoutCache?: boolean;
     bufferPages?: boolean;
-    tableLayouts?: TableLayout;
+    tableLayouts?: {[key: string]: TableLayout};
```
Comment (jeralm) [inline types/pdfmake/interfaces.d.ts:456]: I believe this should be CustomTableLayout instead of TableLayout

File: types/pdfmake/test/pdfmake-module-server-tests.ts [hunk 3]
```diff
@@ -20,10 +21,17 @@ const dd: TDocumentDefinitions = {
...
+        hLineColor: 'red',
+        vLineColor: () => 'red',
+    }
+};
+
 const options: BufferOptions = {
     fontLayoutCache: true,
     bufferPages: true,
-    tableLayouts: 'noBorders',
+    tableLayouts: customTableLayouts,
```
Comment (jeralm) [inline types/pdfmake/test/pdfmake-module-server-tests.ts:34]: Please don't change existing tests that are correct, rather add new tests to cover as much of the new definitions as possible.

**Review by jeralm — CHANGES_REQUESTED** [review]
You should add tests more to cover your changes.
Ideally, this PR should adress everything from #43758 . I'll gladly help if you add me as a collaborator.

File: types/pdfmake/interfaces.d.ts (hunk 1 above)
Comment (syffs) [reply types/pdfmake/interfaces.d.ts:111]: I could not find it in the documentation either, but here is the [related code section](https://github.com/bpampuch/pdfmake/blob/35f3ea19918f7a3b04aa54ad25e94db0c1bd5dac/src/DocMeasure.js#L634)

File: types/pdfmake/test/pdfmake-module-server-tests.ts (hunk 3 above)
Comment (syffs) [reply types/pdfmake/test/pdfmake-module-server-tests.ts:34]: from my understanding, even though it's passing, I don't think that this test is correct: tableLayouts should be an object containing user-defined layouts, providing a pre-defined layout is pointless.
what do you think ?

Comment (jeralm) [reply types/pdfmake/test/pdfmake-module-server-tests.ts:34]: Oh you're right. I didn't notice this was in BufferOptions. Good catch

File: types/pdfmake/interfaces.d.ts (hunk 2 above)
Comment (syffs) [reply types/pdfmake/interfaces.d.ts:456]: @jeralm you're right, I'll update this.

File: types/pdfmake/interfaces.d.ts (hunk 1 above)
Comment (jeralm) [reply types/pdfmake/interfaces.d.ts:111]: I don't know, it looks to me this was done to support the dynamic nature of JS and the default 'auto' value, set right before. If we could use Size | Size[], I'd be more for it, but numbers are not allowed.

In the [playground](http://pdfmake.org/playground.html)
```javascript
dd = { content: { table: {
  widths: 50,
  body: [
    ['Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae'],
    ['Sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae']
  ]
}}};
```
gives an error and
```javascript
dd = { content: { table: {
  widths: '50',
  body: [
    ['Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae'],
    ['Sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae']
  ]
}}};
```
looks buggy.

Percentages, `'*'` and `'auto'` work fine. Any other string seem to give errors.

What do you think?

I think it'd be better to wait until pdmake supports Size, to avoid errors.

Comment (syffs) [reply types/pdfmake/interfaces.d.ts:111]: As you mentioned '*' and 'auto' works fine, so having widths bound to `Size[]` type is limiting. 

I agree that `string | Size[]` is not restrictive enough.... We could do something like this instead:
```
widths?:  '*' | 'auto' | Size[];
```

Comment (jeralm) [reply types/pdfmake/interfaces.d.ts:111]: Sounds like a good compromise.

**Comment by syffs** [comment]
> I'll gladly help if you add me as a collaborator.

done

> Ideally, this PR should adress everything from #43758 . 

sure, the only point from #43758 that I haven't tried to cover is the vfs/font thing, because it needs some digging.

**Comment by jeralm** [comment]
I'll add tests and look up the vfs/font thing

File: types/pdfmake/interfaces.d.ts [hunk 4]
```diff
@@ -79,8 +79,8 @@ export interface CustomTableLayout {
     hLineStyle?: DynamicLayout<LineStyle>;
     vLineStyle?: DynamicLayout<LineStyle>;
     fillColor?: string | DynamicLayout<string>;
-    paddingLeft?: number | DynamicLayout<number>;
-    paddingRight?: number | DynamicLayout<number>;
+    paddingLeft?: DynamicLayout<number>;
+    paddingRight?: DynamicLayout<number>;
```
Comment (jeralm) [inline types/pdfmake/interfaces.d.ts:83]: I haven't been able to use custom table layouts at all in the playground. Why did you change this?

Comment (syffs) [reply types/pdfmake/interfaces.d.ts:83]: I tested it in my env, they both generates exceptions like `layout.paddingLeft is not a function`

Comment (jeralm) [reply types/pdfmake/interfaces.d.ts:83]: Then that means other layout paramters might be erroneous.
Could you try these layouts as well? This might give us some better information.
```typescript
const customTableLayouts: CustomTableLayout[] = [
    {},
    { hLineWidth: () => 50 },
    { vLineWidth: () => 50 },
    { hLineColor: () => '#ff0000' },
    { hLineColor: '#ff0000' },
    { vLineColor: () => '#ff0000' },
    { vLineColor: '#ff0000' },
    { hLineStyle: () => ({ dash: { length: 10, space: 5 }}) },
    { vLineStyle: () => ({ dash: { length: 10, space: 5 }}) },
    { fillColor: () => '#ff0000' },
    { fillColor: '#ff0000' },
    { paddingLeft: () => 50 },
    { paddingRight: () => 50 },
    { paddingTop: () => 50 },
    { paddingBottom: () => 50 },
    { fillOpacity: () => 50 },
    { fillOpacity: 50 },
    { defaultBorder: true }
];
```typescript

Comment (syffs) [reply types/pdfmake/interfaces.d.ts:83]: @jeralm all the layout settings you mentioned seem to work fine !

**Comment by syffs** [comment]
> I'll add tests and look up the vfs/font thing

@jeralm any progress ?

**Comment by jeralm** [comment]
@syffs I guess you didn't notice, I submitted a PR to your fork so you could review and merge my changes.

**Comment by syffs** [comment]
> @syffs I guess you didn't notice, I submitted a PR to your fork so you could review and merge my changes.

@jeralm indeed, i did not :sweat_smile:, sorry about that ! I've just reviewed and merged it

**Review by jeralm — APPROVED** [review]
